File: euclidean-algorithm/euclidean_algorithm.js
function gcd(a, b) {
    while (b !== 0) {
        [a, b] = [b, a % b];
    }
    return a;
}

function gcdRecursive(a, b) {
    if (b === 0) return a;
    return gcdRecursive(b, a % b);
}

function lcm(a, b) {
    return Math.abs(a * b) / gcd(a, b);
}

function extendedGcd(a, b) {
    if (a === 0) {
        return [b, 0, 1];
    }
    
    const [gcdVal, x1, y1] = extendedGcd(b % a, a);
    const x = y1 - Math.floor(b / a) * x1;
    const y = x1;
    
    return [gcdVal, x, y];
}

// Test
console.log(`GCD(48, 18) = ${gcd(48, 18)}`); // 6
console.log(`LCM(48, 18) = ${lcm(48, 18)}`); // 144

File: euclidean-algorithm/EuclideanAlgorithm.kt
fun gcd(a: Int, b: Int): Int {
    var x = a
    var y = b
    while (y != 0) {
        val temp = y
        y = x % y
        x = temp
    }
    return x
}

fun gcdRecursive(a: Int, b: Int): Int {
    return if (b == 0) a else gcdRecursive(b, a % b)
}

fun lcm(a: Int, b: Int): Int {
    return (a * b) / gcd(a, b)
}

fun extendedGcd(a: Int, b: Int): Triple<Int, Int, Int> {
    if (b == 0) {
        return Triple(a, 1, 0)
    }
    val (gcd, x1, y1) = extendedGcd(b, a % b)
    val x = y1
    val y = x1 - (a / b) * y1
    return Triple(gcd, x, y)
}

fun main() {
    println("GCD: ${gcd(48, 18)}")
    println("GCD recursive: ${gcdRecursive(48, 18)}")
    println("LCM: ${lcm(48, 18)}")
    
    val (gcd, x, y) = extendedGcd(48, 18)
    println("Extended GCD: gcd=$gcd, x=$x, y=$y")
}